### ISSUE TITLE
*Chart and DataSeries standardization

### ISSUE THREAD

**Issue by MatthewHerbst** [esbullington/react-d3#219]
**tl;dr**: we need to choose a single format for creating Charts and their elements via the `DataSeries`.

Currently, *Charts and their respective DataSeries are implemented one of two ways:

1) The *Chart creates a `dataSeriesArray` by mapping over the data. Each `DataSeries` produces one chart element, such as a single line (with its circles) in `LineChart`

Ex. from `LineChart`:

```javascript
...
render() { 
    var dataSeriesArray = data.map( (series, idx) => {
        return (
            <DataSeries
              key={idx}
              structure={structure}
              xScale={scales.xScale}
              yScale={scales.yScale}
              seriesName={series.name}
              data={series.values}
              width={innerWidth}
              height={innerHeight}
              fill={props.colors(props.colorAccessor(series, idx))}
              circleRadius={props.circleRadius}
              xAccessor={props.xAccessor}
              yAccessor={props.yAccessor}
              interpolationType={interpolationType}
              displayDataPoints={props.displayDataPoints}
            />
        );
    });

    return (
        <Chart
            ...
        >
            <g
                {dataSeriesArray}
            >
            ...
        </Chart>
    );
}
```

2) The *Chart has a `DataSeries` element which creates all the chart elements. The single `DataSeries` creates all the elements by mapping over them, such as all the bars in `BarChart`.

Ex. from `BarChart`:

```javascript
...
render() {
    return (
        <Chart
            viewBox={props.viewBox}
            legend={props.legend}
            data={props.data}
            margins={props.margins}
            colors={props.colors}
            colorAccessor={props.colorAccessor}
            width={props.width}
            height={props.height}
            title={props.title}
        >
            <g transform={trans} className='rd3-barchart'>
                <DataSeries
                    values={values}
                    labels={labels}
                    yScale={yScale}
                    xScale={yScale}
                    margins={margins}
                    data={props.data}
                    width={props.width - sideMargins}
                    height={props.height - topBottomMargins}
                    colors={props.colors}
                    colorAccessor={props.colorAccessor}
                    hoverAnimation={props.hoverAnimation}
                />
                ...
            </g>
        </Chart>
     );
}
```

We should chose one way of doing things to be standard. This will allow for much simpler code reading, more generic code plugins such as through Mixins, and easier maintenance.

Personally, I prefer the first option. To me, each `DataSeries` should infer a single series of data such as a line or a bar, rather than all of the series of data.

@esbullington @yang-wei

**Comment by yang-wei** [comment]
It's true that we are doing this in 2 ways. As you know our graphs are categorized into 2 types

 * multi-series chart
 * single-series chart

IMO sharable code between these 2 series do not seems that much for me.

> each DataSeries should infer a single series of data such as a line or a bar,

In our `BarChart`, we only have single series and they are rendering `rect` element only. So I think it's ok to keep in this way. In future when we need to support multi series `BarChart`, then we should do it in the way `LineChart` doing.

Sorry for the late reply @MatthewHerbst .

**Comment by MatthewHerbst** [comment]
@yang-wei imho: the way our barchart is implemented right now, it can be trivially used as a multi-series chart, where each series simply has a single data point (each bar is then its own series).

I think all charts should be expressed as multi-series charts. If the user only puts a single series onto them, that is their choice.

As an example of where code differs, the way `props.colors` and `props.colorAccessor` is implemented to create the `fill` for `LineChart` and `BarChart` varies, since in one you pass the `fill` directly to the `DataSeries`, while in the other, you pass the two props themselves. This not only affects maintenance, but also affects things such as which class needs to have default props for required props and so on.

**Comment by yang-wei** [comment]
> it can be trivially used as a multi-series chart,

You get me right ! Let's do this. So the charts we need to refactor are

 * [ ] BarChart
 * [ ] PieChart
 * [ ] Treemap

**Comment by yang-wei** [comment]
I realized that it's not that easy to do it for all the charts especially `scatterchart` and `linechart`.

As you can see in my latest [pr](https://github.com/yang-wei/react-d3/commit/948cb7d22e622c49418eecff5f07719c1dd6cf90) for `linechart`. Our approach now is to flat all data into a new data and map over it to generate voronoi and respective svg component(`rect`, `circle` etc). To do this, the second option you showed above is way more easier.

If you have a greater approach of my latest [pr](https://github.com/yang-wei/react-d3/commit/948cb7d22e622c49418eecff5f07719c1dd6cf90) do let me know.

**Comment by MatthewHerbst** [comment]
Why is it easier? I think the only main difference between the two ways is which component you do that map in. I think the first way is more expressive of what is actually going on. The second way hides that the `DataSeries` is doing much more than just operating on a single `DataSeries`.

Maybe you can explain it better to me, but I really don't understand why we need any of the voronoi stuff. Why are event handlers that simply call `this.setState({things you want to change for animation})` not good enough?

**Comment by yang-wei** [comment]
> but I really don't understand why we need any of the voronoi stuff.

If we get rid of voronoi surely it would be easier.  Voronoi is needed to do better animation in graph like `scatter chart` etc. Looks at the examples of scatter chart in [react-d3-component](http://codesuki.github.io/react-d3-components/example.html). You'd have to hover on the top of circle to trigger tooltip. Voronoi helps to create greater experience.

By implementing voronoi, we need to rely it to handler our events(`click` etc). So our multiple series graph data have to be flatted. We probably won't dump voronoi.

**Comment by MatthewHerbst** [comment]
What's so bad about having to hover over the circle? I'm not sure there is a better way to do it - if you are showing all data points in the y-axis when hovering, what happens if I drop 1,000+ elements into the graph? That's not unreasonable depending on what the chart is showing. 

imho, we should dump animation, unless we want to implement `d3.transition()`

**Comment by yang-wei** [comment]
So I've decided to go with the second option where we only have one `DataSeries`. Maybe `DataSeries` is not the correct name in this case but I think it's fine for now since it's not exposed as our API.

Also for future reference, I write out what is the job of each layer. Our charts are not 100% as stated in below but as you say let's make a standard.

| Component        | What they do           |  
| ------------- |:-------------| 
| ChartName | User input, manipulate data and pass to DataSeries, Axes, Chart etc |
| Chart     | Graph with or without legend, legend styling etc | 
| Data Series      | Apply d3 function (such as voronoi, layout.treemap ) and pass to container via mapping etc  |  
| Container | Managing state for animation and pass prop to children    |   
| SVG Component | Should be stateless, can be reused via other charts |

**Comment by MatthewHerbst** [comment]
One thing I don't understand is the need for containers - why can't say, a Circle, just animate itself?

**Comment by yang-wei** [comment]
@MatthewHerbst [This article](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) explains well =)

**Comment by MatthewHerbst** [comment]
:+1:

**Comment by yang-wei** [comment]
Close due to inactive. Please open a new issue to make further discussion.
